### PR TITLE
fix(AriaCorrection): fixed several aria violations.

### PR DESCRIPTION
**Pull request by jameslau**
fixes: #613

**ISSUE:**
Below are the four listed issues from ticket #613. The first, second, and fourth violation items has been corrected. But as for the third item on the list, further investigation is needed in order to figure out how to dynamically match IDs within divs to `aria-labels` (`aria-label="something"`).

**UPDATE:**
@jgiardino has suggested a fix, https://github.com/patternfly/patternfly-next/issues/613#issuecomment-471068001

**ARIA VIOLATIONS:**
1.  /components/AboutModalBox/examples-full/
```
1 error found in: /components/AboutModalBox/examples-full/

1) Buttons must have discernible text -- https://dequeuniversity.com/rules/axe/3.0/button-name -- impact: critical
      
-------html violation instances-------

<button class="pf-c-button pf-m-plain">
            <i class="fas fa-times" aria-hidden="true"></i></button>
```

2. /components/Button/examples-full/
```
1 error found in: /components/Button/examples-full/

1) Buttons must have discernible text -- https://dequeuniversity.com/rules/axe/3.0/button-name -- impact: critical
      
-------html violation instances-------

<button class="pf-c-button pf-m-plain" disabled="">
    <i class="fas fa-times" aria-hidden="true"></i></button> 
```

3. /components/Toolbar/examples-full/
```
2 errors found in: /components/Toolbar/examples-full/

1) Elements must only use allowed ARIA attributes -- https://dequeuniversity.com/rules/axe/3.0/aria-allowed-attr -- impact: critical
2) Buttons must have discernible text -- https://dequeuniversity.com/rules/axe/3.0/button-name -- impact: critical
      
-------html violation instances-------

<button class="pf-c-options-menu__menu-item pf-m-selected" aria-selected="true">
                  Date
                  <i class="fas fa-check pf-c-options-menu__menu-item-icon" aria-hidden="true"></i>
                </button>
<button class="pf-c-options-menu__menu-item pf-m-selected" aria-selected="true">
                  Ascending
                  <i class="fas fa-check pf-c-options-menu__menu-item-icon" aria-hidden="true"></i>
                </button>
<button class="pf-c-options-menu__menu-item pf-m-selected" aria-selected="true">
                    Option 2
                    <i class="fas fa-check pf-c-options-menu__menu-item-icon" aria-hidden="true"></i>
                  </button>
<button class="pf-c-dropdown__toggle pf-m-plain" id="toolbar-simple-example-dropdown-collapsed-button" aria-expanded="false">
      <i class="fas fa-ellipsis-v" aria-hidden="true"></i></button>
<button class="pf-c-dropdown__toggle pf-m-plain" id="toolbar-pagination-example-dropdown-button" aria-expanded="false">
      <i class="fas fa-ellipsis-v" aria-hidden="true"></i></button>
<button class="pf-c-dropdown__toggle pf-m-plain" id="toolbar-expanded-example-dropdown-simple-button" aria-expanded="true">
      <i class="fas fa-ellipsis-v" aria-hidden="true"></i></button>
<button class="pf-c-dropdown__toggle pf-m-plain" id="toolbar-filter-expanded-example-dropdown-button" aria-expanded="false">
      <i class="fas fa-ellipsis-v" aria-hidden="true"></i></button>
<button class="pf-c-dropdown__toggle pf-m-plain" id="toolbar-filter-expanded-checked-example-dropdown-simple-button" aria-expanded="false">
      <i class="fas fa-ellipsis-v" aria-hidden="true"></i></button>
<button class="pf-c-dropdown__toggle pf-m-plain" id="toolbar-mobile-filter-example-dropdown-button" aria-expanded="false">
      <i class="fas fa-ellipsis-v" aria-hidden="true"></i></button>
<button class="pf-c-dropdown__toggle pf-m-plain" id="toolbar-mobile-filter-checked-example-dropdown-simple-button" aria-expanded="false">
      <i class="fas fa-ellipsis-v" aria-hidden="true"></i></button> 
```

4. /upgrade-examples/AlertUpgradeExamples/examples-full/
```
1 error found in: /upgrade-examples/AlertUpgradeExamples/examples-full/

1) Buttons must have discernible text -- https://dequeuniversity.com/rules/axe/3.0/button-name -- impact: critical
      
-------html violation instances-------

<button class="btn btn-link dropdown-toggle" type="button" id="dropdownKebabRight" data-toggle="dropdown" aria-haspopup="true" aria-expanded="true">
        <span class="fa fa-ellipsis-v"></span>
      </button>
```

### CODE REVIEW

File: src/patternfly/components/Button/button-icon.hbs [hunk 1]
```diff
@@ -1,5 +1,4 @@
 <i class="fas{{#if button-icon--type}} fa-{{button-icon--type}}{{/if}}{{#if button-icon--modifier}} {{button-icon--modifier}}{{/if}}"
-  aria-hidden="true"
```
Comment (jgiardino) [inline src/patternfly/components/Button/button-icon.hbs]: This change isn't needed. `aria-hidden="true" should remain on the `<i>` element. This attribute hides the icon from screen readers, because the sr-accessible text is included on the `<button>` element instead using `aria-label`.

File: src/patternfly/components/Button/button.hbs [hunk 2]
```diff
@@ -1,4 +1,4 @@
-<button class="pf-c-button{{#if button--modifier}} {{button--modifier}}{{/if}}"
+<button class="pf-c-button{{#if button--modifier}} {{button--modifier}}{{/if}}" aria-hidden="true"
```
Comment (jgiardino) [inline src/patternfly/components/Button/button.hbs]: This change isn't needed. See the previous comment.

The issue that was flagged for the Button page is the very last button in the [Button States example](https://pf4.patternfly.org/components/Button/examples/):
```
<button class="pf-c-button pf-m-plain" disabled>
  <i class="fas fa-times" aria-hidden="true"></i>
</button>
```
In this example, the `aria-label="Remove"` is missing. You shouldn't need to update the handlebar files to fix this, because it's already defined for the other examples that are similar. Just this one instance was missing it.

File: src/patternfly/components/Button/examples/button-states-example.hbs [hunk 3]
```diff
@@ -68,14 +68,14 @@
...
 {{/button}}
 {{#> button button--modifier="pf-m-plain pf-m-focus" button--attribute='aria-label="Remove"'}}
-  {{> button-icon button-icon--type="times"}}
+  {{> button-icon button-icon--type="times" button--attribute='aria-label="Remove"'}}
 {{/button}}
 {{#> button button--modifier="pf-m-plain pf-m-active" button--attribute='aria-label="Remove"'}}
-  {{> button-icon button-icon--type="times"}}
+  {{> button-icon button-icon--type="times" button--attribute='aria-label="Remove"'}}
 {{/button}}
 {{#> button button--modifier="pf-m-plain" button--attribute="disabled" button--attribute='aria-label="Remove"'}}
```
Comment (jgiardino) [inline src/patternfly/components/Button/examples/button-states-example.hbs]: All of the changes made so far in this file should be removed. 

I _think_ the issue is here on this line, we have `button--attribute` defined twice, and it looks like it's ignoring the second one where `aria-label` is defined. I'm not sure how to concatenate attributes here. Maybe the following would work?:
`button--attribute='disabled aria-label="Remove"'`
I that doesn't work, then maybe @mcoker can provide a suggestion.

File: src/patternfly/components/Button/examples/button-types-example.hbs [hunk 4]
```diff
@@ -15,5 +15,5 @@
     Link Button
 {{/button}}
 {{#> button button--modifier="pf-m-plain" button--attribute='aria-label="Remove"'}}
-    {{> button-icon button-icon--type="times"}}
+    {{> button-icon button-icon--type="times" button--attribute='aria-label="Remove"'}}
```
Comment (jgiardino) [inline src/patternfly/components/Button/examples/button-types-example.hbs]: Please remove this update.

File: src/patternfly/upgrade-examples/AlertUpgradeExamples/examples/alert-upgrade-example-pf3-notification.hbs [hunk 5]
```diff
@@ -24,8 +24,8 @@
   </div>
   <br>
   <div class="toast-pf alert alert-info">
-    <div class="dropdown pull-right dropdown-kebab-pf">
-      <button class="btn btn-link dropdown-toggle" type="button" id="dropdownKebabRight" data-toggle="dropdown" aria-haspopup="true" aria-expanded="true">
+    <div class="dropdown pull-right dropdown-kebab-pf" aria-label="dropdown toggle">
```
Comment (jgiardino) [inline src/patternfly/upgrade-examples/AlertUpgradeExamples/examples/alert-upgrade-example-pf3-notification.hbs]: `aria-label` should be removed here. 

A `<div>` generally doesn't support `aria-label` unless it has an explicitly assigned role that does support it (usually roles that are interactive elements like buttons or landmark roles).

File: src/patternfly/upgrade-examples/AlertUpgradeExamples/examples/alert-upgrade-example-pf3-notification.hbs [hunk 6]
```diff
@@ -24,8 +24,8 @@
...
+      <button class="btn btn-link dropdown-toggle" type="button" id="dropdownKebabRight" data-toggle="dropdown" aria-haspopup="true" aria-expanded="true" aria-label="dropdown toggle">
```
Comment (jgiardino) [inline src/patternfly/upgrade-examples/AlertUpgradeExamples/examples/alert-upgrade-example-pf3-notification.hbs]: Could you change this to `aria-label="Actions"`?

I'm on the fence about updating the PF3 examples on this page. There will be cases where we cannot fix violations that get flagged. But I'm not opposed to following best practice where possible.

**Comment by jgiardino** [comment]
@jameslau Thanks for working on these updates! I included some comments inline. Let me know if you have questions about these.

File: src/patternfly/components/Button/examples/button-types-example.hbs (hunk 4 above)
Comment (jameslau) [reply src/patternfly/components/Button/examples/button-types-example.hbs]: update has been removed.

File: src/patternfly/upgrade-examples/AlertUpgradeExamples/examples/alert-upgrade-example-pf3-notification.hbs (hunk 6 above)
Comment (jameslau) [reply src/patternfly/upgrade-examples/AlertUpgradeExamples/examples/alert-upgrade-example-pf3-notification.hbs]: `aria-label` changed

File: src/patternfly/upgrade-examples/AlertUpgradeExamples/examples/alert-upgrade-example-pf3-notification.hbs (hunk 5 above)
Comment (jameslau) [reply src/patternfly/upgrade-examples/AlertUpgradeExamples/examples/alert-upgrade-example-pf3-notification.hbs]: `aria-label` removed.

File: src/patternfly/components/Button/examples/button-states-example.hbs (hunk 3 above)
Comment (jameslau) [reply src/patternfly/components/Button/examples/button-states-example.hbs]: Spoke with @mcoker about the consolidating the `disabled` `button--attribute` together with the `aria-label` and this is the result. Not really sure if the `disabled` actually works.

Screen Shot 2019-03-13 at 11 37 08 AM

File: src/patternfly/components/Button/button.hbs (hunk 2 above)
Comment (jameslau) [reply src/patternfly/components/Button/button.hbs]: The last button within the States examples now has `aria-label="Remove"`.

File: src/patternfly/components/AboutModalBox/examples/about-modal-box-example.hbs [hunk 7]
```diff
@@ -4,7 +4,7 @@
     {{/about-modal-box-brand-image}}
   {{/about-modal-box-brand}}
   {{#> about-modal-box-close about-modal-box-close--attribute='aria-label="Close Dialog"'}}
-    {{#> button button--modifier="pf-m-plain"}}
+    {{#> button button--attribute='aria-label="Close Dialog"' button--modifier="pf-m-plain"}}
```
Comment (jgiardino) [inline src/patternfly/components/AboutModalBox/examples/about-modal-box-example.hbs]: I missed this in my original review. Can you make a couple of updates here:

- Currently, the element `div.pf-c-about-modal-box__close` includes `aria-label="Close Dialog"`, which isn't needed. I think you can remove it by removing `about-modal-box-close--attribute='aria-label="Close Dialog"'` from `about-modal-box-close`
- Our [capitalization guidelines](https://v2.patternfly.org/design-guidelines/content/grammar-and-terminology#capitalization) for PF4 uses sentence style capitalization. Can you change "Close Dialog" on the button to "Close dialog"

Comment (jameslau) [reply src/patternfly/components/AboutModalBox/examples/about-modal-box-example.hbs]: Thanks for pointing this out @jgiardino! Corrections has been added.

File: src/patternfly/components/Button/examples/button-states-example.hbs [hunk 8]
```diff
@@ -76,6 +76,6 @@
 {{#> button button--modifier="pf-m-plain pf-m-active" button--attribute='aria-label="Remove"'}}
   {{> button-icon button-icon--type="times"}}
 {{/button}}
-{{#> button button--modifier="pf-m-plain" button--attribute="disabled" button--attribute='aria-label="Remove"'}}
-  {{> button-icon button-icon--type="times"}}
+{{#> button button--modifier="pf-m-plain" button--attribute='disabled aria-label="Remove"'}}
+  {{> button-icon button-icon--type="times" aria-hidden="true"}}
```
Comment (jgiardino) [inline src/patternfly/components/Button/examples/button-states-example.hbs]: The update on line 79 is great! But we don't need to add `aria-hidden="true"` here. It's already included as part of the `button-icon` handlebar.

File: src/patternfly/components/Button/button.hbs (hunk 2 above)
Comment (jgiardino) [inline src/patternfly/components/Button/button.hbs]: Can you revert the change for this file so that the PR shows no changes to this file? Your previous edits are still showing up.

File: src/patternfly/components/Button/button-icon.hbs [hunk 9]
```diff
@@ -1,5 +1,6 @@
-<i class="fas{{#if button-icon--type}} fa-{{button-icon--type}}{{/if}}{{#if button-icon--modifier}} {{button-icon--modifier}}{{/if}}"
-  aria-hidden="true"
+<i class="fas{{#if button-icon--type}} fa-{{button-icon--type}}{{/if}}{{#if button-icon--modifier}} {{button-icon--modifier}}{{/if}}" 
+aria-hidden="true"
   {{#if button-icon--attribute}}
     {{{button-icon--attribute}}}
-  {{/if}}></i>
\ No newline at end of file
+  {{/if}}>
+</i>
```
Comment (jgiardino) [inline src/patternfly/components/Button/button-icon.hbs]: Can you revert the change for this file so that the PR shows no changes to this file?

File: src/patternfly/components/Button/examples/button-types-example.hbs (hunk 4 above)
Comment (jgiardino) [inline src/patternfly/components/Button/examples/button-types-example.hbs]: Can you revert the change for this file so that the PR shows no changes to this file?

**Comment by jgiardino** [comment]
Thanks for making updates, @jameslau! 
I included one comment inline about another update needed.  
It looks like some of the files you edited that didn't need edits weren't reverted properly, and they're still showing up in this PR. I noted these files with inline comments, too. Reach out if you need help figuring out how to do this (I'm assuming it's easy, but have never done it myself 🙃). 

Also, `/components/Toolbar/examples-full/` was noted as part of issue https://github.com/patternfly/patternfly-next/issues/613. Were you planning to handle those updates in this PR or a separate PR?

**Comment by jameslau** [comment]
Hi @jgiardino, thanks for taking another look at the PR. The `/components/Toolbar/examples-full/` is actually covered in all three tickets #613, #617, #619. I just pushed a PR up for #617 (https://github.com/patternfly/patternfly-next/pull/1575) which may cover for most of the violations? I hope...

In terms of the "revert" issue, I'll go look at that right now and see how to adjust them back.

File: src/patternfly/components/Button/examples/button-states-example.hbs (hunk 8 above)
Comment (jameslau) [reply src/patternfly/components/Button/examples/button-states-example.hbs]: `aria-hidden="true"` has been removed.

File: src/patternfly/components/Button/button.hbs (hunk 2 above)
Comment (jameslau) [reply src/patternfly/components/Button/button.hbs]: file has been reverted back.

File: src/patternfly/components/Button/button-icon.hbs (hunk 9 above)
Comment (jameslau) [reply src/patternfly/components/Button/button-icon.hbs]: This file has been reverted back as well.

File: src/patternfly/components/Button/examples/button-types-example.hbs (hunk 4 above)
Comment (jameslau) [reply src/patternfly/components/Button/examples/button-types-example.hbs]: Reverted the code back for this one as well.

**Comment by srambach** [comment]
This changes the HTML structure of a PF3 upgrade example, which isn't the point that we work with existing PF3 HTML? Also, Dana's PR to exclude updated examples was merged this morning. So would we want to leave that as is?

**Comment by jgiardino** [comment]
I'm on the fence about that one. I know in our PF3 examples, we don't include aria-labels for things like kebabs, but it doesn't alter the html structure or visual appearance to do so. I don't think we should make major modifications to PF3 examples to fix issues, because that's a lot of work, but if it's a single attribute that makes something accessible when before it wasn't, then I think it's ok. 🤷‍♀️

**Comment by srambach** [comment]
@jgiardino, then I'm fine with it if you are.
